### PR TITLE
CodeMirror: Add keyboard navigation for precise code intelligence

### PR DESCRIPTION
**Pull request by umpox**
https://user-images.githubusercontent.com/9516420/195355312-355d038c-8d75-4aff-a347-dda1ffab9bc8.mp4

## App preview:

- [Web](https://sg-web-tr-blob-accessibility.onrender.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.02% (+0.52 kb) | 0.06% (+9.20 kb) | 0.07% (+8.68 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/tr/blob-accessibility/compare-report.html;tab=live_object) for a full comparison between the commits 6c65131 and 7836dc2 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: client/shared/src/settings/settings.ts [hunk 1]
```diff
@@ -20,6 +20,7 @@ export interface IClient {
 export interface Settings {
     extensions?: { [extensionID: string]: boolean }
     'codeIntel.referencesPanel'?: 'tree-view' | 'tabbed'
+    'codeIntel.keyboardNavigation'?: 'none' | 'token'
```
Comment (umpox) [inline client/shared/src/settings/settings.ts:23]: More specific name? How about `advancedKeyboardNavigation`?

`intelligentKeyboardNavigation`

File: client/web/src/repo/blob/CodeMirrorBlob.tsx [hunk 2]
```diff
@@ -58,6 +61,9 @@ const staticExtensions: Extension = [
         '.selected-line': {
             backgroundColor: 'var(--code-selection-bg)',
         },
+        '.selected-line:focus': {
+            boxShadow: 'none',
+        },
```
Comment (umpox) [inline client/web/src/repo/blob/CodeMirrorBlob.tsx:66]: This removes the focus outline we have. We already have the background-color so I think focus can be implied here. In a follow up PR I'm going to add <kbd>↑</kbd> <kbd>↓</kbd> controls that allows users to quickly move focus across the doc and an alternative "focused line" style.

File: client/web/src/repo/blob/backend.ts [hunk 3]
```diff
@@ -119,3 +161,88 @@ export const usePrefetchBlobFormat = (): HighlightResponseFormat => {
...
+                if (!data.repository?.commit) {
+                    throw new Error('Commit not found')
+                }
+
+                if (!data.repository.commit.blob?.lsif) {
+                    return null
+                }
+
+                // If we have multiple definitions we just return the first one.
+                return data.repository.commit.blob.lsif.definitions.nodes[0]
```
Comment (umpox) [inline client/web/src/repo/blob/backend.ts:241]: This appears to be how the current go-to-def works, I'd be open for discussions about what is actually best to do here though, open the references panel?

**Comment by umpox** [comment]
Closed in favour of https://github.com/sourcegraph/sourcegraph/pull/42936 and https://github.com/sourcegraph/sourcegraph/pull/42937

**Pull request closed without merging**